File: BFS_DFS_etc_Data_structure/all_of_graph/graph-max-island-area.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int R2, C2;
char** Grid;
bool** visited2;
int dr4[4] = {1, -1, 0, 0};
int dc4[4] = {0, 0, 1, -1};

typedef struct {
    int r, c;
} Node;

int dfs_area_stack(int sr, int sc){
    Node* stack = (Node*)malloc(R2*C2*sizeof(Node));
    int top = -1;
    top++;
    stack[top].r = sr;
    stack[top].c = sc;
    visited2[sr][sc] = true;
    int area = 0;
    
    while(top >= 0){
        Node curr = stack[top--];
        area++;
        int rr = curr.r, cc = curr.c;
        for(int i=0; i<4; i++){
            int nr = rr + dr4[i];
            int nc = cc + dc4[i];
            if(nr>=0 && nr<R2 && nc>=0 && nc<C2){
                if(Grid[nr][nc] == '1' && !visited2[nr][nc]){
                    visited2[nr][nc] = true;
                    top++;
                    stack[top].r = nr;
                    stack[top].c = nc;
                }
            }
        }
    }
    free(stack);
    return area;
}

int max_area_of_island(char** M, int rows, int cols){
    R2 = rows;
    C2 = cols;
    Grid = M;
    
    visited2 = (bool**)malloc(R2*sizeof(bool*));
    for(int i=0; i<R2; i++){
        visited2[i] = (bool*)calloc(C2, sizeof(bool));
    }
    
    int max_area = 0;
    
    for(int r=0; r<R2; r++){
        for(int c=0; c<C2; c++){
            if(Grid[r][c] == '1' && !visited2[r][c]){
                int area = dfs_area_stack(r, c);
                if(area > max_area){
                    max_area = area;
                }
            }
        }
    }
    
    // cleanup
    for(int i=0; i<R2; i++){
        free(visited2[i]);
    }
    free(visited2);
    return max_area;
}

int main(void){
    int rows = 5, cols = 5;
    char* sample[5] = {
        "11000",
        "11001",
        "00011",
        "00000",
        "10110"
    };
    
    char** M = (char**)malloc(rows*sizeof(char*));
    for(int i=0; i<rows; i++){
        M[i] = sample[i];
    }
    
    int ans = max_area_of_island(M, rows, cols);
    printf("Max area of island: %d\n", ans);
    
    free(M);
    return 0;
}

File: BFS_DFS_etc_Data_structure/all_of_graph/graph-max-island-area.py
"""
Given a 2D grid M of '1' and '0', find the maximum area (number of '1' cells) among 
all islands, where islands are connected 4-directionally. 
If there are no '1' cells at all, return 0.
"""
from typing import List

def max_area_of_island(M: List[List[str]]) -> int:
    if not M or not M[0]:
        return 0
    rows, cols = len(M), len(M[0])
    visited = [[False]*cols for _ in range(rows)]
    directions = [(1,0),(-1,0),(0,1),(0,-1)]
    
    def dfs_area(r, c):
        stack = [(r,c)]
        visited[r][c] = True
        area = 0
        while stack:
            rr, cc = stack.pop()
            area += 1
            for dr, dc in directions:
                nr, nc = rr + dr, cc + dc
                if 0 <= nr < rows and 0 <= nc < cols:
                    if M[nr][nc] == '1' and not visited[nr][nc]:
                        visited[nr][nc] = True
                        stack.append((nr, nc))
        return area
    
    max_area = 0
    for r in range(rows):
        for c in range(cols):
            if M[r][c] == '1' and not visited[r][c]:
                area = dfs_area(r, c)
                if area > max_area:
                    max_area = area
    return max_area

if __name__ == "__main__":
    grid = [
        ['1','1','0','0','0'],
        ['1','1','0','0','1'],
        ['0','0','0','1','1'],
        ['0','0','0','0','0'],
        ['1','0','1','1','0']
    ]
    ans = max_area_of_island(grid)
    print("Max area of island:", ans)
    # The exact area depends on how the islands connect. 
    # E.g., the top-left cluster is area=4, 
    # top-right cluster is area=2, 
    # bottom-left cluster might be area=1, 
    # bottom row (4,2),(4,3) => 2 together, etc.
